### PR TITLE
Add `PrimaryEntityDimensionPairs` validation rule

### LINKED ISSUE #103: Add ValidationRule ensuring there are no duplicate primary entity parings
# Context / Problem

1. Multiple `SemanticModel` objects can have the same primary `Entity`
2. `Dimension` names are not required to unique across `SemanitcModel`s in the `SemanticManifest`

Together statements 1 and 2 mean that it is possible for `SemanticModel`s to have the same primary `Entity` and have `Dimension`s of the same name. This is problematic because if a dimension `<primary_entity>__<dimension_nam>` is specified in a `WhereFilter`, it can be ambiguous which dimension is actually being referenced.

# Solution

Disallow `SemanticModel`s with the same primary `Entity` to have duplicate `Dimension`s

# Accetance Criteria

A validation rule exists which disallows `SemanticModel`s with the same primary `Entity` to have duplicate `Dimension`s

### PR DESCRIPTION
**Pull request by QMalcolm**
Resolves #103

### Description

Adds a `PrimaryEntityDimensionPairs` validation rule which validates that there are no duplicate primary entity + dimension pairings across the semantic manifest.

### Checklist

- [X] I have read [the contributing guide](https://github.com/dbt-labs/dbt-semantic-interfaces/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [X] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [X] This PR includes tests, or tests are not required/relevant for this PR
- [X] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-semantic-interfaces/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: tests/fixtures/semantic_manifest_yamls/simple_semantic_manifest/semantic_models/user_sm_source.yaml
Comment (tlento) [inline tests/fixtures/semantic_manifest_yamls/simple_semantic_manifest/semantic_models/user_sm_source.yaml:14]: Not that it matters much but I'd rather rename this to like `ds_source_date` or something.

**Review by tlento — APPROVED** [review]
LGTM, thanks!

**Comment by QMalcolm** [comment]
Gonna wait to merge this after #124 is merged

File: dbt_semantic_interfaces/validations/unique_valid_name.py
Comment (QMalcolm) [reply dbt_semantic_interfaces/validations/unique_valid_name.py:252]: Yea I've spent some time thinking about dependent rules but it'd require quite the overhaul. The simplest solution is to add a `ValidationWarning` when detected

**Pull request merged**
